src/core/model.jl

```diff
@@ -190,13 +194,73 @@ Return an iterator over all agents of the model.
 """
 allagents(model) = values(model.agents)
 
-
 """
     allids(model)
 Return an iterator over all agent IDs of the model.
 """
 allids(model) = keys(model.agents)
 
+#######################################################################################
+# %% Higher order collections
+#######################################################################################
+export group_agent_iter, group_id_iter, group_agent_map, group_id_map
+
+"""
+    group_id_iter(order::Int, model::ABM)
+
+Return an iterator over all agent IDs of the model, group by order. When `order = 2`, the
+iterator returns agent pairs, and when `order = 3`: agent triples.
+
+See also [`group_agent_iter`](@ref).
+"""
+group_id_iter(order::Int, model::ABM) =
+    Iterators.product((by_id(model) for _ in 1:order)...)
+
+"""
+    group_agent_iter(order::Int, model::ABM)
+
+Return an iterator over all agents of the model, group by order. When `order = 2`, the
+iterator returns ID pairs, e.g `(1, 2)` and when `order = 3`: ID triples, e.g.
+`(1, 7, 8)`.
+
+See also [`group_id_iter`](@ref).
+"""
+group_agent_iter(order::Int, model::ABM) =
+    Iterators.product((map(i -> model[i], by_id(model)) for _ in 1:order)...)
+
+"""
+    group_id_map(order::Int, f::Function, model::ABM)
+    group_id_map(order::Int, f::Function, model::ABM, filter::Function)
+
+Applies function `f` to all grouped agent IDs of a [`group_id_iter`](@ref). `f` must take
+the form `f(model, a, b)`, where the number of arguments (after `model`) is equal to
+`order`.
```

Comment:
The syntax `f(model, a, b)` is not clear to me, when followed by "where the number of arguments..". Does this refer to `a, b`? I.e. if order was 3 you would need `a, b, c`? Then it is better to define the call as `f(model, args...)` with `args...` as many as `order`.

Reply:
Yep, that's correct. It's essentially a curried function.